task/move.go

```diff
@@ -86,27 +87,82 @@ func Move(fetch Fetcher, put Putter) {
 	}
 }
 
+func (o *oldStore) fetchB() (thing Thing, ok bool, err error) {
+	// simulate the delay
+	time.Sleep(3 * time.Second)
+	fmt.Println("fetchB: fetching...", o.bookNo)
+
+	if o.bookNo == 5 {
+		return thing, ok, errors.New("Fetch encountered network issues")
+	}
+
+	if o.bookNo > len(o.bookInventory)-1 {
+
+		return thing, ok, err
+	}
+
+	thing, ok = o.bookInventory[o.bookNo], true
+	o.bookNo++
+	return
+}
+
+func (n *newStore) putB(thing Thing) error {
+	time.Sleep(3 * time.Second)
+
+	n.inventory = append(n.inventory, thing)
+
+	if len(n.inventory) == 5 {
+		return errors.New("store is filled up")
+	}
+
+	return nil
+}
+
 // MaybeMove is exactly the same as Move except that it may return an error
 // because fetch() and put() may return errors. If no errors occur then
 // MaybeMove returns under the same conditions as Move(). If an error occurs
 // then MaybeMove returns earlier even if there are more Things to fetch().
 func MaybeMove(fetch MaybeFetcher, put MaybePutter) error {
-	// [TODO]
+	ch := make(chan Thing)
+	errCh := make(chan error, 2)
+	quit := make(chan struct{})
+	defer close(errCh)
 
 	go func() {
-		// [TODO]
-		t, ok, err := fetch()
-		_, _, _ = t, ok, err // [Remove later]
-		// [TODO]
+		defer close(ch)
+
+		for {
+			select {
+			// get a signal to stop the goroutine
+			case <-quit:
+				fmt.Println("Terminated from Parent Goroutine...")
+				return
+			default:
+				t, ok, err := fetch()
+				if err != nil {
+					fmt.Println("Fetch encountered error")
+					errCh <- err
+					return
+				}
+
+				if !ok {
+					return
+				}
+
+				ch <- t
+			}
+		}
 	}()
 
-	var t Thing // [Remove later]
-	// [TODO]
-	err := put(t)
-	_ = err // [Remove later]
-	// [TODO]
+	for thing := range ch {
+		if err := put(thing); err != nil {
+			fmt.Println("Error in adding to new store")
+			close(quit) // signal to end the fetch goroutine
+			errCh <- err
+		}
+	}
 
-	return nil // [Remove later]
+	return <-errCh
```

Comment:
What happens now if there are no errors? Nothing will ever get sent on `errCh` so this will block forever.

Reply:
Thanks for the feedback, looking at it, I closed the channel before sending the error that got placed on it, what this results in as before is that only the first reported error is sent from the channel.